### ISSUE TITLE
[quantization] Make range expansion (including zero) configurable in observer

### ISSUE THREAD

**Issue by mhs4670go** [Samsung/TICO#637]
## Summary

The recent fix for asymmetric quantization updates the observed range to always include zero when the original range does not cross zero. While this resolves the inconsistency between `scale` and `zero_point` (e.g., clamping issues), it also changes the quantization behavior and may reduce precision in certain cases.

This issue proposes making this behavior configurable.

Related: #634

## Background

In asymmetric quantization, when the observed range does not include zero (e.g., strictly positive or strictly negative values), computing `zero_point` from the original range can result in values outside the representable integer range, leading to clamping and inconsistent quantization.

To address this, the current implementation expands the range to include zero:
- if `min > 0`, set `min = 0`
- if `max < 0`, set `max = 0`

This ensures a valid `(scale, zero_point)` pair but introduces a trade-off.

## Problem

Expanding the range to include zero can significantly increase the scale when the original range is far from zero.

Example:
- Original range: `[1024, 1030]`
- Adjusted range: `[0, 1030]`
- Result: much larger scale → reduced quantization resolution

This may negatively impact accuracy for tensors with large offsets.

## Proposal

Introduce a configuration option to control this behavior.

### Option 1: Boolean flag

```python
include_zero_in_range: bool = True
```

### Option 2: Policy-based configuration

```python
asym_range_policy: Literal[
    "force_include_zero",  # current behavior
    "keep_original"        # do not modify observed range
]
```

###  Additional Considerations

- Emit a warning when the observed range does not include zero, especially when range expansion is applied.
- Consider extending support for more advanced heuristics (e.g., conditionally expanding only when the range is close to zero).

### Expected Outcome

- Provide users with control over the trade-off between numerical stability and quantization precision.
- Improve transparency and debuggability of quantization behavior.

**Comment by Torrero** [comment]
In my opinion, option 2 is preferable, since the behavior of each configuration can be described in a convenient way, and, in addition, in the future the policy can be expanded, for example, by automatically selecting behavior, depending on the estimation of the quantization error.

**Comment by dvsav** [comment]
I doubt there's an option to not expand the variable range to include zero.
I've written a script that illustrates what happens when we don't apply range expansion.
It plots the dependency graph between the fake-quantized value and the original (unquantized) floating-point value.
```python
import torch
import matplotlib.pyplot as plt
import numpy as np

def calc_scale_zp(
    q_min: int,
    q_max: int,
    fp_min: float,
    fp_max: float,
) -> tuple[float, int]:
    """Calculates scale and zero point"""
    rng = fp_max - fp_min

    scale = torch.tensor(rng / (q_max - q_min))
    zp = torch.round(q_min - fp_min / scale).clamp(q_min, q_max).to(torch.int)

    return scale.item(), zp.item()

def fake_quantize(
    x: float,
    scale: float,
    zp: int,
    q_min: int,
    q_max: int,
) -> float:
    return torch.fake_quantize_per_tensor_affine(
        torch.tensor(x),
        scale=torch.tensor(scale),
        zero_point=torch.tensor(zp),
        quant_min=q_min,
        quant_max=q_max,
    ).item()

def plot_asymmetric_quantization(
    fp_min: float,
    fp_max: float,
    q_min: int,
    q_max: int,
    image_path: str,
):
    """Generate and plot the quantization distortion for asymmetric ranges"""

    # Calculate quantization parameters
    scale, zp = calc_scale_zp(q_min, q_max, fp_min, fp_max)

    # Generate test values uniformly distributed in fp range
    num_samples = 1000
    fp_values = np.linspace(fp_min, fp_max, num_samples)

    # Quantize and dequantize all values
    quantized_values = []
    for x in fp_values:
        q_val = fake_quantize(x, scale, zp, q_min, q_max)
        quantized_values.append(q_val)

    # Create plot
    fig, ax = plt.subplots(figsize=(12, 6))

    # Plot original vs quantized values
    ax.plot(fp_values, fp_values, 'b-', alpha=0.3, label='Original Values', linewidth=2)
    ax.plot(fp_values, np.array(quantized_values), 'r-', label='Quantized Values', linewidth=1.5)

    # Add theoretical boundaries
    y_max = scale * (q_max - zp)
    y_min = scale * (q_min - zp)
    ax.axhline(y=y_max, color='green', alpha=0.5, linestyle=':', label=f'Theoretical Max ({y_max:.2f})')
    ax.axhline(y=y_min, color='orange', alpha=0.5, linestyle=':', label=f'Theoretical Min ({y_min:.2f})')

    # Add reference lines
    ax.axvline(x=fp_min, color='gray', linestyle=':', alpha=0.5)
    ax.axvline(x=fp_max, color='gray', linestyle=':', alpha=0.5)

    # Formatting
    ax.set_xlabel('Floating Point Value (Input)', fontsize=12)
    ax.set_ylabel('Value After Quantization/Dequantization', fontsize=12)
    ax.set_title(f"Asymmetric Quantization Distortion\n"
                 f"Scale={scale:.6f}, Zero Point={zp}\n"
                 f"Range: [{fp_min}, {fp_max}] → [{q_min}, {q_max}]", 
                 fontsize=14, fontweight='bold')

    ax.legend(loc='upper right')
    ax.grid(True, alpha=0.3)
    ax.set_xlim(min(0, fp_min), max(0, fp_max))   # Set horizontal range
    ax.set_ylim(min(0, fp_min), max(0, fp_max))   # Set vertical range
    ax.set_aspect('equal')

    plt.tight_layout()
    plt.xlim(min(0, fp_min), max(0, fp_max))
    plt.ylim(min(0, fp_min), max(0, fp_max))
    plt.savefig(image_path)

def parse_arguments():
    import argparse

    parser = argparse.ArgumentParser(
        description="Fake Quantization Plotter."
    )

    parser.add_argument(
        "--fp_min",
        type=float,
        default=1.0,
        help="Floating-point value minimum.",
    )

    parser.add_argument(
        "--fp_max",
        type=float,
        default=3.55,
        help="Floating-point value maximum.",
    )

    parser.add_argument(
        "--q_min",
        type=int,
        default=0,
        help="Quantized value minimum.",
    )

    parser.add_argument(
        "--q_max",
        type=int,
        default=255,
        help="Quantized value maximum.",
    )

    parser.add_argument(
        "--image_path",
        type=str,
        default="image.png",
        help="Image file name.",
    )
    
    return parser.parse_args()
    

if __name__ == "__main__":
    args = parse_arguments()
    plot_asymmetric_quantization(
        fp_min=args.fp_min,
        fp_max=args.fp_max,
        q_min=args.q_min,
        q_max=args.q_max,
        image_path=args.image_path,
    )
```

Unfortunately, I can't post images to GitHub for some reason.
Yet, if you run the script as shown below, you'll see that the issue becomes more and more serious as `fp_min` approaches `fp_max / 2`.

```bash
python quant.py \
    --fp_min=1.0 \
    --fp_max=3.55 \
    --q_min=0 \
    --q_max=255 \
    --image_path="image_1.0-3.55_0-255.png"
```

```bash
python quant.py \
    --fp_min=1.0 \
    --fp_max=2.536 \
    --q_min=0 \
    --q_max=255 \
    --image_path="image_1.0-2.536_0-255.png"
```

```bash
python quant.py \
    --fp_min=1.0 \
    --fp_max=2.024 \
    --q_min=0 \
    --q_max=255 \
    --image_path="image_1.0-2.024_0-255.png"
```

If we don't expand the range, we get all fake-quantized values clamped to `(0, fp_max-fp_min)` range. For example, if we try to quantize (without range expansion) the value ranging from `1.0` to `2.0`, any values greater than `1.0` get clamped to `1.0` - and thus, as you may guess, *all* the values in the range `1.0...2.0` are clamped to `1.0`.

What do you think?

**Comment by mhs4670go** [comment]
Good point — without range expansion, quantization can collapse due to zero-point clamping. I agree this is not just a precision trade-off but a correctness issue.

That said, it also suggests that the core problem may lie in how the observed range is defined. Instead of relying purely on min/max, it could be beneficial to consider distribution-aware approaches such as percentile-based range estimation (e.g., clipping extreme outliers). This could help keep the range closer to the bulk of the data, reduce excessive scale expansion, and in some cases naturally include zero without requiring explicit adjustment.

We could explore this as a complementary direction alongside the current fix.

Thank you for your comments again!
